### ISSUE TITLE
v3 Performance Issue with workflow-instances Endpoint

### ISSUE THREAD

**Issue by rolandgriesser** [elsa-workflows/elsa-core#4350]
The performance of the /workflow-instances-Endpoint is really slow due to EF Core including the `Data` column in the query, although it won't be mapped to the model and returned later on.

The log output of the executed query:
```
SELECT [w].[Id], [w].[CorrelationId], [w].[CreatedAt], [w].[Data], [w].[DefinitionId], [w].[DefinitionVersionId], [w].[FinishedAt], [w].[Name], [w].[Status], [w].[SubStatus], [w].[UpdatedAt], [w].[Version]
      FROM [Elsa].[WorkflowInstances] AS [w]
      ORDER BY [w].[CreatedAt] DESC
      OFFSET @__p_0 ROWS FETCH NEXT @__p_1 ROWS ONLY
``` 

I think this is related to using a static projection method (`WorkflowInstanceSummary.FromInstance`), which EF Core can't translate. If I change the `.Select` to directly create the constructor of `WorkflowInstanceSummary` in the following line, the performance of calling the endpoint increases from 2,4 minutes! to a couple of milliseconds.

https://github.com/elsa-workflows/elsa-core/blob/1e7f9543116a0caabf3cf26d30d72b512ec1740d/src/modules/Elsa.EntityFrameworkCore/Modules/Management/WorkflowInstanceStore.cs#L75

What's a bit strange: If I try to reproduce this outside of the Elsa packages, it still includes the `Data` column in the query, but the performance is way better. Is there any logging/change tracking activated, that would decrease the performance?

```
var context = new ManagementElsaDbContext(new DbContextOptionsBuilder<ManagementElsaDbContext>().UseSqlServer("Server=...").LogTo(i => Console.WriteLine(i)).Options);

            var set = context.WorkflowInstances;
            WorkflowInstanceFilter filter = new();
            var o = new WorkflowInstanceOrder<DateTimeOffset>
            {
                KeySelector = p => p.CreatedAt,
                Direction = Elsa.Common.Entities.OrderDirection.Descending
            };
            var queryable = Filter(set.AsQueryable(), filter).OrderBy(o);
            var count = await queryable.LongCountAsync();
            queryable = queryable.Paginate(new PageArgs() { Limit = 15, Offset = 9 });
            var entities = await queryable.Select(x => Elsa.Workflows.Management.Models.WorkflowInstanceSummary.FromInstance(x)).ToListAsync();
```

**Comment by sfmskywalker** [comment]
Hello @rolandgriesser,

Thank you for bringing this to our attention.

Regarding the performance difference between using a static method invocation and explicitly constructing a new object shape in EF Core, we've delved into the matter. It appears that when using a static factory method for projection, EF Core cannot directly translate the method's internals into SQL. As a result, it defaults to selecting all columns from the table, even if the method only uses a subset of them. On the other hand, when using the `new` keyword directly in the LINQ query, EF Core can optimize the SQL query to select only the necessary columns. Given this understanding, we will address the issue by instantiating `WorkflowInstanceSummary` using the `new` keyword to ensure optimized SQL generation.

As for the performance difference you observed, we believe it might be related to the deserialization process. When using the Elsa packages, after fetching the data from the database, the `Data` column contents are deserialized into objects. This deserialization process can introduce a noticeable overhead, especially if the `Data` column contains complex or large serialized objects. This might explain the additional time taken when using the Elsa packages compared to directly querying the DB context.

We appreciate your feedback and will work on implementing the necessary optimizations to enhance performance. Please let us know if you have any further insights or questions.

**Comment by rolandgriesser** [comment]
Thanks for the feedback!
In the meantime, I'm wrote an registered an adapter around the `EFCoreWorkflowInstanceStore` where I reimplemented the affected methods, using the following approach - which basically works the same as your approach, but can be optimized by EF Core. It still looks clean and is reusable (I had to use the `Extension` class as `WorkflowInstanceSummary` isn't partial): https://stackoverflow.com/a/73272146

```
 public async Task<Page<WorkflowInstanceSummary>> SummarizeManyAsync<TOrderBy>(WorkflowInstanceFilter filter, PageArgs pageArgs, WorkflowInstanceOrder<TOrderBy> order, CancellationToken cancellationToken = default)
        {
            await using var dbContext = await _store.CreateDbContextAsync(cancellationToken);
            var set = dbContext.WorkflowInstances;
            var queryable = Filter(set.AsQueryable(), filter).OrderBy(order);
            var count = await queryable.LongCountAsync(cancellationToken);
            queryable = queryable.Paginate(pageArgs);
            var entities = await queryable.Select(WorkflowInstanceSummaryExtensions.FromInstance()).ToListAsync(cancellationToken);

            return Page.Of(entities, count);
        }


    internal static class WorkflowInstanceSummaryExtensions
    {
        public static Expression<Func<WorkflowInstance, WorkflowInstanceSummary>> FromInstance()
            => workflowInstance => new WorkflowInstanceSummary
            {
                Id = workflowInstance.Id,
                DefinitionId = workflowInstance.DefinitionId,
                DefinitionVersionId = workflowInstance.DefinitionVersionId,
                Version = workflowInstance.Version,
                Status = workflowInstance.Status,
                SubStatus = workflowInstance.SubStatus,
                CorrelationId = workflowInstance.CorrelationId,
                Name = workflowInstance.Name,
                CreatedAt = workflowInstance.CreatedAt,
                UpdatedAt = workflowInstance.UpdatedAt,
                FinishedAt = workflowInstance.FinishedAt
            };
    }
```

**Comment by sfmskywalker** [comment]
Fix via #4376 - many thanks @rolandgriesser for your contributions!
